README.md
Postcode.nl API Magento 2 Plugin by Experius
=============

Magento 2 port of Magento 1 extension [Github](https://github.com/postcode-nl/PostcodeNl_Api_MagentoPlugin).

Look up and validate Dutch addresses at Checkout with this Magento extension from [Postcode.nl](https://www.postcode.nl/services/adresdata/webshops).

- Reduce address errors
- Reduce costs from undelivered mail and packages
- Offer extra service for your customers with easy address entry
- Improve sale conversion

Data
=============

This plugin uses the [Postcode.nl Address API](https://www.postcode.nl/services/adresdata/webshops) to validate and enrich Dutch address data.
You can obtain an account for the Postcode.nl API [here](https://account.postcode.nl/Main/register).


CHANGELOG.md
## 1.11.2 (2023-10-30)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.11.2)

*  [BUGFIX][IN23-254] Avoid using closing slash with void tags. *(Boris van Katwijk)*


## 1.11.1 (2023-07-12)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.11.1)

*  Update LicenceCheck.php *(mhaagen85)*


## 1.11.0 (2022-10-17)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.11.0)

*  [BUGFIX][PWAI-716] Avoid unused variables; see details below. *(Boris van Katwijk)*


## 1.10.0 (2022-09-06)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.10.0)

*  Changed function because ReadOnly Class can not be used in PHP8.1 *(Demi Holland)*
*  Edit trim function because PHP8.1 compatibility *(Demi Holland)*


## 1.9.0 (2021-07-15)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.9.0)

*  [FEATURE][DRRS-47] - Updated postcode api url with new url to keep continued support *(Rens Wolters)*


## 1.8.2 (2021-05-12)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.8.2)

*  [BUGFIX] - Fix housenumber addition *(Ruben Panis)*


## 1.8.1 (2021-02-01)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.8.1)

*  [BUGFIX] Solved "TypeError: Cannot read property 'get' of undefined" when country is set to Belgium *(Mr. Lewis)*


## 1.8.0 (2020-10-28)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.8.0)

*  [FEATURE] Introduce pluggable function to alter sort order of the postcode fields for different checkouts than the default Magento OnePage. *(Boris van Katwijk)*


## 1.7.0 (2020-10-05)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.7.0)

*  [FEATURE]Added pattern to housenumber input for numeric keyboard on ios *(joeydankbaar)*


## 1.6.0 (2020-10-02)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.6.0)

*  [DOCS] Updated the CHANGELOG.md *(Lewis Voncken)*
*  [FEATURE] Split GraphQl functionality from module *(René Schep)*


## 1.5.0 (2020-10-01)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.5.0)

*  [FEATURE] Changed the template for the housenumber to set input type number *(Lewis Voncken)*


## 1.4.2 (2020-09-30)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.4.2)

*  Update URLs, services.postcode.nl no longer exists *(SjonHortensius)*
*  Prevent javascript error by adding a warning comment when you want to use seperate address fields *(joeydankbaar)*
*  [BUGFIX] Fixed issue where street gets [object] in name. *(Egor Dmitriev)*
*  [BUGFIX] Fixed issue where street gets duplicate values while changing housenumber, additon or on initial load *(Egor Dmitriev)*


## 1.4.1 (2020-09-22)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.4.1)

*  [FEATURE] Updated GraphQl Exception type for lookupAddress errors *(René Schep)*


## 1.4.0 (2020-09-15)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.4.0)

*  [DOC] Cleaned some code *(René Schep)*
*  [FEATURE] Added support for GraphQl *(René Schep)*
*  [FEATURE] Added magento version restriction *(René Schep)*
*  [BUGFIX] Typo *(René Schep)*
*  [BUGFIX] Fixed version restraint *(René Schep)*


## 1.3.4 (2020-04-09)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.3.4)

*  [BUGFIX] removed region_id set, expects an int not a string. *(Experius)*


## 1.3.3 (2019-11-27)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.3.3)

*  Prevent notice showing undefined *(jordy2607)*


## 1.3.2 (2019-11-21)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.3.2)

*  [BUGFIX] solved invalid width for the input fields in the checkout *(Mr. Lewis)*


## 1.3.1 (2019-07-01)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.3.1)

*  [TASK] - Change echo json_encode data to return resultJsonFactory as to prevent pipeline error *(Cas Satter)*


## 1.3.0 (2019-01-29)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.3.0)

*  Typo *(Michiel Gerritsen)*
*  [TASK] Added Api keycheck in backend and system message when key is invalid *(Matthijs Breed)*
*  [BUGFIX] Fixed checkout postcode validation in checkout form to fill in magento housenumber field when user switches to manual input *(Matthijs Breed)*
*  [TASK] Improved readability of javascript *(Hexmage)*
*  [TASK] Fixed formatting *(Hexmage)*
*  Fixed Formatting *(Hexmage)*
*  PSR-2 Formatting *(Hexmage)*
*  PSR-2 Formatting *(Hexmage)*
*  PSR-2 Formatting *(Hexmage)*
*  Fixed Formatting *(Hexmage)*
*  PSR-2 Formatting *(Hexmage)*


## 1.2.1 (2018-08-15)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.2.1)

*  [BUGFIX] Fixed Housenumber addition not being seperated to next street field *(René Schep)*


## 1.2.0 (2018-03-13)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.2.0)

*  [TASK] Updated the Styling in the Checkout and Added a notice with 'Fill in your Shipping Address below:' *(Lewis Voncken)*


## 1.1.6 (2018-03-13)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.1.6)

*  [TASK] Added canRestore to the configuration fields *(Lewis Voncken)*


## 1.1.5 (2018-03-13)

[View Release](git@github.com:experius/Magento-2-Module-Experius-Postcode-NL.git/commits/tag/1.1.5)


Block/Checkout/LayoutProcessor.php
<?php
/**
 * A Magento 2 module named Experius Postcode
 * Copyright (C) 2017 Experius
 *
 * This file is part of Experius Postcode.
 *
 * Experius Postcode is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

namespace Experius\Postcode\Block\Checkout;

use Magento\Checkout\Block\Checkout\LayoutProcessorInterface;
use Magento\Framework\View\Element\AbstractBlock;

class LayoutProcessor extends AbstractBlock implements LayoutProcessorInterface
{
    protected $scopeConfig;

    protected $logger;

    protected $changeFieldPositions = false;

    public function __construct(
        \Magento\Framework\View\Element\Template\Context $context,
        array $data = []
    ) {
        $this->scopeConfig = $context->getScopeConfig();
        $this->logger = $context->getLogger();

        parent::__construct($context, $data);
    }

    /**
     * Process js Layout of block
     *
     * @param array $result ($jsLayout)
     * @return array
     */
    public function process($result)
    {
        if (
            $this->scopeConfig->getValue(
                'postcodenl_api/general/enabled',
                \Magento\Store\Model\ScopeInterface::SCOPE_STORE
            )
            && isset($result['components']['checkout']['children']['steps']['children']['shipping-step']['children']
                ['shippingAddress']['children']['shipping-address-fieldset'])
        ) {
            $this->changeFieldPositions = $this->scopeConfig->getValue(
                'postcodenl_api/advanced_config/change_sort_order',
                \Magento\Store\Model\ScopeInterface::SCOPE_STORE
            );

            $shippingFields = $result['components']['checkout']['children']['steps']['children']
            ['shipping-step']['children']['shippingAddress']['children']
            ['shipping-address-fieldset']['children'];

            $shippingFields = array_merge($shippingFields, $this->getPostcodeFieldSet('shippingAddress', 'shipping'));

            if ($this->changeFieldPositions) {
                $shippingFields = $this->changeAddressFieldPosition($shippingFields);
            }

            $shippingFields = $this->addClasses('shippingAddress', $shippingFields);

            $result['components']['checkout']['children']['steps']['children']
            ['shipping-step']['children']['shippingAddress']['children']
            ['shipping-address-fieldset']['children'] = $shippingFields;

            $result = $this->getBillingFormFields($result);
        }
        return $result;
    }

    /**
     * Get billing form fields
     *
     * @param array $result
     * @return array
     */
    public function getBillingFormFields($result)
    {
        if (isset(
            $result['components']['checkout']['children']['steps']['children']
            ['billing-step']['children']['payment']['children']['payments-list']
        )) {
            $paymentForms = $result['components']['checkout']['children']['steps']['children']
            ['billing-step']['children']['payment']['children']
            ['payments-list']['children'];

            foreach (array_keys($paymentForms) as $paymentMethodForm) {
                $paymentMethodCode = str_replace('-form', '', $paymentMethodForm);

                if (!isset($result['components']['checkout']['children']['steps']['children']['billing-step']
                    ['children']['payment']['children']['payments-list']['children'][$paymentMethodCode . '-form'])) {
                    continue;
                }

                $billingFields = $result['components']['checkout']['children']['steps']['children']
                ['billing-step']['children']['payment']['children']
                ['payments-list']['children'][$paymentMethodCode . '-form']['children']['form-fields']['children'];

                $billingPostcodeFields = $this->getPostcodeFieldSet('billingAddress' . $paymentMethodCode, 'billing');

                $billingFields = array_merge($billingFields, $billingPostcodeFields);

                $billingFields = $this->addClasses('billingAddress' . $paymentMethodCode, $billingFields);

                if ($this->changeFieldPositions) {
                    $billingFields = $this->changeAddressFieldPosition($billingFields);
                }

                $result['components']['checkout']['children']['steps']['children']['billing-step']
                ['children']['payment']['children']['payments-list']['children'][$paymentMethodCode . '-form']
                ['children']['form-fields']['children'] = $billingFields;
            }
        }

        return $result;
    }

    /**
     * Get postcode fieldset
     *
     * @param $scope
     * @param $addressType
     * @return array[]
     */
    public function getPostcodeFieldSet($scope, $addressType)
    {
        return [
            'experius_postcode_fieldset' => [
                'component' => 'Experius_Postcode/js/view/form/postcode',
                'type' => 'group',
                'config' => [
                    "customScope" => $scope,
                    "template" => 'Experius_Postcode/form/group',
                    "additionalClasses" => "experius_postcode_fieldset",
                    "loaderImageHref" => $this->getViewFileUrl('images/loader-1.gif')
                ],
                'sortOrder' => (string)($this->getStartSortOrder() + 5),
                'children' => $this->getPostcodeFields($scope, $addressType),
                'provider' => 'checkoutProvider',
                'addressType' => $addressType,
            ]
        ];
    }

    /**
     * Get postcode fields
     *
     * @param $scope
     * @param $addressType
     * @return array[]
     */
    public function getPostcodeFields($scope, $addressType)
    {
        $postcodeFields =
            [
                'experius_postcode_postcode' => [
                    'component' => 'Magento_Ui/js/form/element/abstract',
                    'config' => [
                        "customScope" => $scope,
                        "template" => 'ui/form/field',
                        "elementTmpl" => 'ui/form/element/input',
                    ],
                    'provider' => 'checkoutProvider',
                    'dataScope' => $scope . '.experius_postcode_postcode',
                    'label' => __('Postcode'),
                    'sortOrder' => (string)($this->getStartSortOrder() + 15),
                    'validation' => [
                        'required-entry' => true,
                        'min_text_length' => 6,
                    ]
                ],
                'experius_postcode_housenumber' => [
                    'component' => 'Magento_Ui/js/form/element/abstract',
                    'config' => [
                        "customScope" => $scope,
                        "template" => 'ui/form/field',
                        "elementTmpl" => 'Experius_Postcode/form/element/housenumber'
                    ],
                    'provider' => 'checkoutProvider',
                    'dataScope' => $scope . '.experius_postcode_housenumber',
                    'label' => __('Housenr.'),
                    'sortOrder' => (string)($this->getStartSortOrder() + 25),
                    'validation' => [
                        'required-entry' => true,
                    ],
                ],
                'experius_postcode_housenumber_addition' => [
                    'component' => 'Magento_Ui/js/form/element/select',
                    'config' => [
                        "customScope" => $scope,
                        "template" => 'ui/form/field',
                        "elementTmpl" => 'ui/form/element/select'
                    ],
                    'provider' => 'checkoutProvider',
                    'dataScope' => $scope . '.experius_postcode_housenumber_addition',
                    'label' => __('Addition'),
                    'sortOrder' => (string)($this->getStartSortOrder() + 30),
                    'validation' => [
                        'required-entry' => false,
                    ],
                    'options' => [],
                    'visible' => false,
                ],
                'experius_postcode_housenumber_addition_manual' => [
                    'component' => 'Magento_Ui/js/form/element/abstract',
                    'config' => [
                        "customScope" => $scope,
                        "template" => 'ui/form/field',
                        "elementTmpl" => 'ui/form/element/input'
                    ],
                    'provider' => 'checkoutProvider',
                    'dataScope' => $scope . '.experius_postcode_housenumber_addition_manual',
                    'label' => __('Addition'),
                    'sortOrder' => (string)($this->getStartSortOrder() + 30),
                    'validation' => [
                        'required-entry' => false,
                    ],
                    'options' => [],
                    'visible' => false,
                ],
                'experius_postcode_disable' => [
                    'component' => 'Magento_Ui/js/form/element/abstract',
                    'config' => [
                        "customScope" => $scope,
                        "template" => 'ui/form/field',
                        "elementTmpl" => 'ui/form/element/checkbox'
                    ],
                    'provider' => 'checkoutProvider',
                    'dataScope' => $scope . '.experius_postcode_disable',
                    'description' => __('Enter address manually'),
                    'sortOrder' => (string)($this->getStartSortOrder() + 50),
                    'validation' => [
                        'required-entry' => false,
                    ],
                    'addressType' => $addressType
                ]
            ];

        return $postcodeFields;
    }

    /**
     * Change address fields position(s)
     *
     * @param $addressFields
     * @return mixed
     */
    public function changeAddressFieldPosition($addressFields)
    {
        if (isset($addressFields['country_id'])) {
            $addressFields['country_id']['sortOrder'] = (string)($this->getStartSortOrder());
        }

        // 'experius_postcode_fieldset' is $this->getStartSortOrder() + 5

        if (isset($addressFields['street'])) {
            $addressFields['street']['sortOrder'] = (string)($this->getStartSortOrder() + 10);
        }

        // 'experius_postcode_postcode' field default is + 15

        if (isset($addressFields['city'])) {
            $addressFields['city']['sortOrder'] = (string)($this->getStartSortOrder() + 20);
        }

        // 'experius_postcode_housenumber' is + 25
        // 'experius_postcode_housenumber_addition' and 'experius_postcode_housenumber_addition_manual' are + 30

        if (isset($addressFields['postcode'])) {
            $addressFields['postcode']['sortOrder'] = (string)($this->getStartSortOrder() + 35);
        }

        if (isset($addressFields['region'])) {
            $addressFields['region']['sortOrder'] = (string)($this->getStartSortOrder() + 40);
        }

        if (isset($addressFields['region_id'])) {
            $addressFields['region_id']['sortOrder'] = (string)($this->getStartSortOrder() + 45);
        }

        // 'experius_postcode_disable' is + 50

        return $addressFields;
    }

    /**
     * Get start sort order
     * Hook for plugins to alter it's default: 900
     *
     * @TODO: Introduce system configuration to alter this value
     * @return int
     */
    public function getStartSortOrder()
    {
        return 900;
    }

    /**
     * Add classes to address fields
     *
     * @param $scope
     * @param $addressFields
     * @return mixed
     */
    public function addClasses($scope, $addressFields)
    {
        foreach (['street', 'region_id', 'region', 'country_id', 'city', 'postcode'] as $field) {
            if (isset($addressFields[$field])) {
                $configAdditionalClasses = null;
                $additionalClasses = $scope . '-' . $field;
                if (isset($addressFields[$field]['config']['additionalClasses'])) {
                    $configAdditionalClasses = $addressFields[$field]['config']['additionalClasses'];
                    $additionalClasses = $configAdditionalClasses . ' ' . $additionalClasses;
                }
                $addressFields[$field]['config']['additionalClasses'] = $additionalClasses;
            }
        }

        return $addressFields;
    }
}
